### PR TITLE
Feature/teachers can delete groups

### PR DESCRIPTION
**Pull request by irinahpe**
Enables teachers to delete groups that were created by this plugin.
Fixes issues #33 and #205

### CODE REVIEW

File: locallib.php [hunk 1]
```diff
@@ -965,72 +966,99 @@ public function distrubute_choices() {
     public function synchronize_allocation_and_grouping() {
         require_capability('moodle/course:managegroups', $this->context);
 
-        $groupingidname = ratingallocate_MOD_NAME . '_instid_' . $this->ratingallocateid;
         // Search if there is already a grouping from us.
-        $grouping = groups_get_grouping_by_idnumber($this->course->id, $groupingidname);
-        $groupingid = null;
-        if (!$grouping) {
+        if (!$this->db->record_exists(this_db\ratingallocate_id_grouping::TABLE, ['ratingallocateid' => $this->ratingallocateid])) {
```
Comment (Laur0r) [inline locallib.php:970]: If you have an if-else statement were you want to check whether a record exists and continue working with the existing record, it is better to check by using `if(!$obj=get_record(...))` (see line 155 in lib.php). This way you only have to call the DB once for this and can continue with $obj for the if or else statement.

File: locallib.php [hunk 2]
```diff
@@ -965,72 +966,99 @@ public function distrubute_choices() {
...
-        }
+        // Loop through existing choices.
+        foreach ($choices as $choice) {
+            if ($this->db->record_exists(this_db\ratingallocate_choices::TABLE,
+                    ['id' => $choice->id])) {
 
-        // Dind all associated groups in this grouping.
-        $groups = groups_get_all_groups($this->course->id, 0, $groupingid);
+                // Checks if there is already a group for this choice.
+                if ($this->db->record_exists(this_db\ratingallocate_choice_group::TABLE,
```
Comment (Laur0r) [inline locallib.php:1005]: see above, better use get_record

File: locallib.php [hunk 3]
```diff
@@ -965,72 +966,99 @@ public function distrubute_choices() {
...
         $choices = $this->get_choices_with_allocationcount();
 
-        // Make a new array containing only the identifiers of the choices.
-        $choiceids = array();
-        foreach ($choices as $id => $choice) {
-            $choiceids[$groupidentifierfromchoiceid($choice->id)] = array('key' => $id);
-        }
+        // Loop through existing choices.
+        foreach ($choices as $choice) {
+            if ($this->db->record_exists(this_db\ratingallocate_choices::TABLE,
```
Comment (Laur0r) [inline locallib.php:1001]: This is totally fine as you do not need the choice record later on here.

File: locallib.php [hunk 4]
```diff
@@ -965,72 +966,99 @@ public function distrubute_choices() {
...
-                groups_delete_group($group->id);
-            }
-        }
+                    // Get the group from the choice_group Table.
+                    $groupids = $this->db->get_records_select(this_db\ratingallocate_choice_group::TABLE,
+                        "choiceid=$choice->id",
+                        null,
+                        '',
+                        'id, groupid');
+                    // Only one object in groupids because there should only be one entry in the table with this choiceid.
```
Comment (Laur0r) [inline locallib.php:1014]: If there should be only one object available, is there corresponding error handling for that? Then get_record() should be enough here and there is no need for a loop. But there should be sufficient checks when adding to the database to guarantee there are no duplicate entries for the same choice.

File: locallib.php [hunk 5]
```diff
@@ -965,72 +966,99 @@ public function distrubute_choices() {
...
-            $choiceidentifier = $groupidentifierfromchoiceid($choiceid);
-            groups_add_member($choiceids[$choiceidentifier]['groupid'], $userid);
+
+            // Get the group corresponding to the choiceid.
+            $groupids = $this->db->get_records_select(this_db\ratingallocate_choice_group::TABLE,
+                "choiceid=$choiceid",
+                null,
+                '',
+                'groupid');
+            // Only one object in groupids because there should only be one entry in the table with this choiceid.
```
Comment (Laur0r) [inline locallib.php:1054]: see above, get_records_select and loop not necessary here then

File: db/install.xml [hunk 6]
```diff
@@ -83,5 +83,29 @@
...
+        <FIELD NAME="groupid" TYPE="int" LENGTH="10" NOTNULL="true" SEQUENCE="false" COMMENT="Reference to a group."/>
+        <FIELD NAME="choiceid" TYPE="int" LENGTH="10" NOTNULL="true" SEQUENCE="false" COMMENT="Reference to a choice item."/>
+      </FIELDS>
+      <KEYS>
+        <KEY NAME="primary" TYPE="primary" FIELDS="id"/>
+        <KEY NAME="groupid" TYPE="foreign" FIELDS="groupid" REFTABLE="groups" REFFIELDS="id"/>
+        <KEY NAME="choiceid" TYPE="foreign" FIELDS="choiceid" REFTABLE="ratingallocate_choices" REFFIELDS="id"/>
+      </KEYS>
+    </TABLE>
+    <TABLE NAME="ratingallocate_id_grouping" COMMENT="Maps the Instance of the Allocation with the Grouping.">
```
Comment (Laur0r) [inline db/install.xml:98]: This table should be named ratingallocate_grouping to indicate that it relates the instance with groupings, to be consistent there shouldn't be an id in there

File: db/install.xml [hunk 7]
```diff
@@ -83,5 +83,29 @@
         <KEY NAME="groupid" TYPE="foreign" FIELDS="groupid" REFTABLE="groups" REFFIELDS="id"/>
       </KEYS>
     </TABLE>
+    <TABLE NAME="ratingallocate_choice_group" COMMENT="Maps Groups and the Allocations on which they are created with">
```
Comment (Laur0r) [inline db/install.xml:86]: I struggle with how similar the two tables are called (`ratingallocate_choice_group` vs `ratingallocate_group_choices`). As the other is already used by some users, maybe change this to `ratingallocate_allocationchoice_group`?

File: backup/moodle2/backup_ratingallocate_activity_stepslib.php [hunk 8]
```diff
@@ -69,25 +98,30 @@ protected function define_structure() {
         $ratingallocate_choice->add_child($groupchoices);
         $groupchoices->add_child($groupchoice);
 
-        // Define sources
+        $ratingallocate_choices->add_child($choicegroups);
```
Comment (Laur0r) [inline backup/moodle2/backup_ratingallocate_activity_stepslib.php:101]: I think this should be `$ratingallocate_choice` instead as this corresponds to one choice.
Grouping should also be build in this tree.

**Review by Laur0r — COMMENTED** [review]
Great work! It seems that however the backup and restore process does not take the generated groups into account (these are not restored in new courses), can you have another look at that?

File: db/install.xml (hunk 7 above)
Comment (irinahpe) [reply db/install.xml:86]: It's not possible to change the name to something much different since tablenames can only be 28 characters long

**Pull request closed without merging**